doc/architectural-overview.md

```diff
@@ -137,38 +137,40 @@ Code for dealing with mongodb is found primarily in the `state`, `state/watcher`
 
 ## API
 
-Juju controllers expose an API endpoint over a websocket connection. The methods
-available over the API are broken down by client; there's a `Client` facade that
-exposes the methods used by clients, an `Agent` facade that exposes the methods
-common to all agents, and a wide range of worker-specific *facades* that individually
-deal with particular chunks of functionality implemented by one agent or another
-(for example, `Provisioner`, `Upgrader`, and `Uniter`, each used by the eponymous
+Juju controllers expose an API endpoint over a websocket connection. Related
+API endpoints are grouped into a set called a "facade". Each facade generally
+serves either the client, the controller, or Juju agents. As a general rule,
+each worker that needs to talk to the API server will get its own facade (for
+example, `Provisioner`, `Upgrader`, and `Uniter`, each used by the eponymous
 worker types).
 
-The API server is implemented in the `apiserver` top level package. Each API
-facade has it's own subpackage (e.g. `apiserver/provisioner`). The code under
-`apiserver` is the only code that is allowed to import from the `state`
-package.
+The API server is implemented in the `apiserver` top level package. Facades
+live inside the `apiserver/facades` package. There is a subpackage `agent` for
+agent facades, `client` for client facades, and `controller` for controller
+facades.
 
-Various facades share functionality; for example, the Life method is used by many
-worker facades. In these cases, the method is implemented on a separate type,
-which is embedded in the facade implementation.
+Various facades share functionality; for example, the Life method is used by
+many worker facades. In these cases, the method is implemented on a separate
+type in the `apiserver/common` package, then embedded in the facade.
```

Comment:
> the method is implemented on a separate type in the `apiserver/common` package, then embedded in the facade.

This is an anti-pattern. Anything you expose via embedding is then made public and needs to be explicitly versioned. What you should do, which is important for the document and to document it, is to make the common part a field and expose what you need. Thus, allowing you to version on common methods.

Pseudo code for this would be:

```go
type Facade struct {
     lifer common.Life
}

func (f Facade) Life(ctx context.Context, modelArgs params.LifeArgs) (params.LifeResult, error) {
     return f.lifer.Life(ctx, modelArgs)
}
```

Again, this allows us to move away from Life. You can't do that without breaking all facades that embed common APIs.

I would also note that any public functions on the facade become part of the facade's API.

Reply:
I agree with you, I'm just describing what I've seen in many facades.